kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractExactlyOnceKafkaOutputOperator.java

```diff
@@ -0,0 +1,385 @@
+    Map<Integer,Long> currentOffsets = getCurrentOffsets();
+
+    if (storedOffsets == null || currentOffsets == null) {
+      //TODO: Take some action
+      return;
+    }
+
+    KafkaConsumer consumer = KafkaConsumerInit();
+
+    List<TopicPartition> topicPartitions = new ArrayList<>();
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      topicPartitions.add(new TopicPartition(getTopic(), entry.getKey()));
+    }
+
+    consumer.assign(topicPartitions);
+
+    // From each partitions in a topic
+    // 1. Read the tuples between stored offset and the latest offset
+    // 2. Check the Key to filter out the messages
+    // 3. Store the recovered tuples in the HashSet
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      Long storedOffset = 0L;
+      Integer currentPartition = entry.getKey();
+      Long currentOffset = entry.getValue();
+
+      if (storedOffsets.containsKey(currentPartition)) {
+        storedOffset = storedOffsets.get(currentPartition);
+      }
+
+      if (storedOffset >= currentOffset) {
+        continue;
+      }
+
+      consumer.seek(new TopicPartition(getTopic(), currentPartition), storedOffset);
+
+      int kafkaAttempt = 0;
+
+      while ( true ) {
+
+        ConsumerRecords<String, String> consumerRecords = consumer.poll(100);
+
+        if (consumerRecords.count() == 0) {
+          ++kafkaAttempt;
+
+          if (kafkaAttempt == KAFKA_CONNECT_ATTEMPT) {
+            break;
+          }
+        } else {
+          kafkaAttempt = 0;
+        }
+
+        boolean crossedBoundary = false;
+
+        for (ConsumerRecord consumerRecord : consumerRecords) {
+
+          if ( !doesKeyBelongsThisInstance(operatorId, (String)consumerRecord.key()) ) {
+            continue;
+          }
+
+          recoveredTuples.add((T)consumerRecord.value());
+
+          if ( consumerRecord.offset() >= currentOffset ) {
+            crossedBoundary = true;
+            break;
+          }
+        }
+
+        if ( crossedBoundary ) {
+          break;
+        }
+      }
+    }
+
+    consumer.close();
+  }
+
+  private Map<Integer,Long> getCurrentOffsets()
+  {
+    Map<Integer, Long> currentOffsets = null;
+
+    try {
+      currentOffsets = getPartitionsAndOffsets();
+    } catch (ExecutionException e) {
+      e.printStackTrace();
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    }
+
+    return currentOffsets;
+  }
+
+  private Map<Integer,Long> getStoredOffsets()
+  {
+    Map<Integer,Long> storedOffsets = null;
+    try {
+      storedOffsets = (Map<Integer,Long>)this.windowDataManager.load(operatorId, windowId);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    return storedOffsets;
+  }
+
+  private KafkaConsumer KafkaConsumerInit()
+  {
+    Properties props = new Properties();
+    props.put("bootstrap.servers", getConfigProperties().get("bootstrap.servers"));
```

Comment:
Can you use Java Constants instead of hardcoding property strings?
You can find the public static string constants in org.apache.kafka.clients.producer.ProducerConfig